internal/server/filesystem/server.go

```diff
@@ -2,24 +2,227 @@ package filesystem
 )
 
-// Server is the internal server for the filesytem API group.
-type Server struct{}
+type Server struct {
+	kubeletCSIPluginsPath string
+	kubeletPodPath        string
+	hostAPI               API
+}
+
+const MAX_PATH_LENGTH_WINDOWS = 260
+
+var invalidPathCharsRegexWindows = regexp.MustCompile(`["/\:\?\*|]`)
+
+type API interface {
+	PathExists(path string) (bool, error)
+	Mkdir(path string) error
+	Rmdir(path string) error
+	LinkPath(tgt string, src string) error
+}
+
+func NewServer(hostOS string, kubeletCSIPluginsPath string, kubeletPodPath string, hostAPI API) (*Server, error) {
+	if hostOS != "windows" {
+		return nil, fmt.Errorf("Unsupported OS for FileSystem API server: %s", hostOS)
+	}
+	return &Server{
+		kubeletCSIPluginsPath: kubeletCSIPluginsPath,
+		kubeletPodPath:        kubeletPodPath,
+		hostAPI:               hostAPI,
+	}, nil
+}
+
+func isDriveLetterWindows(c uint8) bool {
+	return ('a' <= c && c <= 'z' || 'A' <= c && c <= 'Z')
+}
+
+func containsInvalidCharactersWindows(path string) bool {
+	if isAbsWindows(path) {
+		path = path[3:]
+	}
+	if invalidPathCharsRegexWindows.MatchString(path) {
+		return true
+	}
+	if strings.Contains(path, `..`) {
+		return true
+	}
+	return false
+}
+
+func isUNCPathWindows(path string) bool {
+	// check for UNC/pipe prefixes like "\\"
+	if len(path) < 2 {
+		return false
+	}
+	if path[0] == '\\' && path[1] == '\\' {
+		return true
+	}
+	return false
+}
+
+func isAbsWindows(path string) bool {
+	// for Windows check for C:\\.. prefix only
+	// UNC prefixes of the form \\ are not considered
+	// absolute in the context of CSI proxy
+	if len(path) < 3 {
+		return false
+	}
+	c := path[0]
+	if !isDriveLetterWindows(c) {
+		return false
+	}
+	if path[1] != ':' {
+		return false
+	}
+	if path[2] != '\\' && path[2] != '/' {
+		return false
+	}
+	return true
+}
+
+func (s *Server) validatePathWindows(pathCtx internal.PathContext, path string) error {
+	prefix := ""
+	if pathCtx == internal.PLUGIN {
+		prefix = s.kubeletCSIPluginsPath
+	} else if pathCtx == internal.CONTAINER {
+		prefix = s.kubeletPodPath
+	} else {
+		return fmt.Errorf("Invalid PathContext: %v", pathCtx)
+	}
+
+	pathlen := len(path)
+
+	if pathlen > MAX_PATH_LENGTH_WINDOWS {
+		return fmt.Errorf("Path length %d exceeds maximum characters: %d", pathlen, MAX_PATH_LENGTH_WINDOWS)
+	}
+
+	if len(path) > 0 && (path[0] == '\\') {
+		return fmt.Errorf("Invalid character \\ at begining of path: %s", path)
+	}
+
+	if isUNCPathWindows(path) {
+		return fmt.Errorf("Unsupported UNC path prefix: %s", path)
+	}
+
+	if containsInvalidCharactersWindows(path) {
+		return fmt.Errorf("Path contains invalid characters: %s", path)
+	}
+
+	if isAbsWindows(path) && !strings.HasPrefix(path, prefix) {
+		return fmt.Errorf("Absolute path: %s is not within context path: %s", path, prefix)
+	}
+
+	return nil
+}
+
+func (s *Server) abs(pathCtx internal.PathContext, path string) (string, error) {
+	if isAbsWindows(path) {
```

Comment:
Shouldn't we let the caller decide what type the path needs to interpreted as - either ABSOLUTE path or PLUGIN or CONTAINER path? Currently, we seem to be have a mode in which we assume that the user intended that its absolute path by looking at the path and then fall through to other modes of PLUGIN and CONTAINER. I am thinking we could either have: 
-  Callers performs the append and passes the full path in all cases.
or 
- Have another mode internal.ABSOLUTE where we can explicitly know the caller intent. We can do the validations we do in isAbsWindows.

@ddebroy - WDYT ?

Reply:
Good point. I was thinking about this along the lines of mkdir/rmdir APIs in golang where the absolute or relative nature of a path is determined by the API implementation rather than having to be explicitly specified by a parameter from the caller. Since the nature of the path is trivial to determine (especially given our scoped support surface i.e. no UNC paths, etc), I was thinking we can do without an extra parameter.

I think under most scenarios CSI node plugins typically deal with absolute paths passed by the in-tree CSI plugin in kubelet as required by the spec (https://github.com/container-storage-interface/spec/blob/master/spec.md#nodepublishvolume and https://github.com/container-storage-interface/spec/blob/master/spec.md#nodestagevolume, etc).

Do you have any specific scenario in mind where there might be ambiguity?

Reply:
Decision is to allow Absolute paths only for now. Relative path support can be added later if it becomes necessary.

Reply:
is the code only allow absolute path?